File: src/main/scala/com/zooplus/jacekb/learningTime/akka/pi/Pi.scala
package com.zooplus.jacekb.learningTime.akka.pi

import akka.actor._
import com.zooplus.jacekb.learningTime.akka.Master
import com.zooplus.jacekb.learningTime.akka.pi.Commons.Calculate

/**
 * Created with IntelliJ IDEA.
 * User: jacek_bilski
 * Date: 29/04/13
 * Time: 08:50
 */
object Pi extends App {

	calculate(nrOfWorkers = 8, nrOfElements = 10000, nrOfMessages = 100000)

	//	class Master(nrOfWorkers: Int,
	//							 nrOfMessages: Int,
	//							 nrOfElements: Int,
	//							 listener: ActorRef) extends Actor {
	//
	//		var pi: BigDecimal = 0
	//		var nrOfResults: Int = _
	//		val start: Long = System.currentTimeMillis
	//
	//		val workerRouter = context.actorOf(
	//			Props[Worker].withRouter(RoundRobinRouter(nrOfWorkers)), name = "workerRouter")
	//
	//		def receive = {
	//			case Calculate ⇒
	//				for (i ← 0 until nrOfMessages)
	//					workerRouter ! Work(i * nrOfElements, nrOfElements)
	//			case Result(value) ⇒
	//				pi += value
	//				nrOfResults += 1
	//				if (nrOfResults == nrOfMessages) {
	//					// Send the result to the listener
	//					listener ! PiApproximation(pi, duration = (System.currentTimeMillis - start).millis)
	//					// Stops this actor and all its supervised children
	//					context.stop(self)
	//				}
	//		}
	//	}

	def calculate(nrOfWorkers: Int, nrOfElements: Int, nrOfMessages: Int) {
		// Create an Akka system
		val system = ActorSystem("PiSystem")

		// create the result listener, which will print the result and
		// shutdown the system
		val listener = system.actorOf(Props[Listener], name = "listener")

		// create the master
		val master = system.actorOf(Props(new Master(
			nrOfWorkers, nrOfMessages, nrOfElements, listener)),
			name = "master")

		// start the calculation
		master.tell(new Calculate, ActorRef.noSender)
	}
}

File: src/main/java/com/zooplus/jacekb/learningTime/akka/pi/java/Pi.java
package com.zooplus.jacekb.learningTime.akka.pi.java;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;

/**
 * Created with IntelliJ IDEA.
 * User: jacek_bilski
 * Date: 17.09.14
 * Time: 09:37
 */
public class Pi {

    public static void main(String[] args) {
        Pi pi = new Pi();
        pi.calculate(8, 10000, 100000);
    }

    private void calculate(int nrOfWorkers, int nrOfElements, int nrOfMessages) {
        ActorSystem system = ActorSystem.create("PiSystem");
        ActorRef listener = system.actorOf(Props.create(Listener.class), "listener");
        ActorRef manager = system.actorOf(Props.create(Manager.class, nrOfWorkers, nrOfMessages, nrOfElements, listener), "manager");
        manager.tell(new Calculate(), ActorRef.noSender());
    }
}
